File: midas.py
import sys
from functools import partial
from threading import Thread
from queue import Queue, Empty
from evdev import InputDevice, ecodes, list_devices


def guess_direction(X, chunk_size=4):
    dir = 0
    for i in range(len(X) - chunk_size):
        seen_lt = False
        seen_gt = False
        for j in range(chunk_size - 1):
            a = X[i+j]
            b = X[i+j+1]
            seen_lt |= b < a
            seen_gt |= b > a
        dir += seen_gt - seen_lt
    return dir


def guess_event(data):
    if not data:
        return
    y_data = []
    x_data = []
    c_data = []
    for t, y in data:
        if   t == 0: y_data.append(y)
        elif t == 1: x_data.append(y)
        else:        c_data.append(y)
    if not c_data:
        return
    dy = guess_direction(y_data)
    dx = guess_direction(x_data)
    c = max(c_data)
    # find out main direction change
    if abs(dx) > abs(dy):
        dir_name = "left" if dx < 0 else "right"
    else:
        dir_name = "up" if dy < 0 else "down"
    key_name = "4"  if c == ecodes.BTN_TOOL_QUADTAP else "3"
    return "%s.%s" % (key_name, dir_name)


def timed(q):
    buffer = []
    while True:
        try:
            buffer.append(q.get(timeout=0.1))
        except Empty:
            evt = guess_event(buffer)
            if evt is not None:
                print(evt, flush=True)
            buffer.clear()


def main(device_path):
    dev = InputDevice(device_path)
    q = Queue()

    Thread(target=partial(timed, q)).start()
    taps = (ecodes.BTN_TOOL_TRIPLETAP,
            ecodes.BTN_TOOL_QUADTAP)

    for event in dev.read_loop():
        type  = event.type
        code  = event.code
        value = event.value
        if type == ecodes.EV_ABS:
            if code == ecodes.ABS_Y:   q.put((0, value))
            elif code == ecodes.ABS_X: q.put((1, value))
        elif type == ecodes.EV_KEY and code in taps:
            q.put((2, code))


if __name__ == '__main__':
    devices = [InputDevice(path) for path in list_devices()]
    if sys.argv[1] == "list":
        for device in devices:
            print(device.path, device.name)
    elif not sys.argv[1].startswith('/'):
        for device in devices:
            if device.name == sys.argv[1]:
                main(device.path)
                break
    else:
        main(sys.argv[1])

File: midas.go
package main

import "os"
import "fmt"
import "time"
import "github.com/desertbit/timer"
import "github.com/gvalkov/golang-evdev"

func gatherEvents(device *evdev.InputDevice) chan *evdev.InputEvent {
	sink := make(chan *evdev.InputEvent)
	go func() {
		for {
			ev, err := device.ReadOne()
			if err != nil {
				sink <- nil
				close(sink)
				break
			}
			sink <- ev
		}
	}()
	return sink
}

func update_buff(buff *[4]int32, value int32, i int) (int, int) {
	if i < 4 {
		buff[i] = value
		return 0, i + 1
	}
	// shift values and recalculate
	buff[0] = buff[1]
	buff[1] = buff[2]
	buff[2] = buff[3]
	buff[3] = value
	asc := buff[0] <= buff[1] && buff[1] <= buff[2] && buff[2] <= buff[3]
	dsc := buff[0] >= buff[1] && buff[1] >= buff[2] && buff[2] >= buff[3]
	if asc && !dsc {
		return +1, i
	} else if !asc && dsc {
		return -1, i
	} else {
		return 0, i
	}
}

func abs(i int) int {
	if i < 0 {
		return -i
	}
	return i
}

func get_gesture_name(dir_x, dir_y int, c_max uint16) (event string, ok bool) {
	switch c_max {
	case evdev.BTN_TOOL_QUADTAP:
		event = "4."
	case evdev.BTN_TOOL_TRIPLETAP:
		event = "3."
	default:
		ok = false
		return
	}
	abs_x := abs(dir_x)
	abs_y := abs(dir_y)
	if abs_x > abs_y && abs_x >= 15 {
		ok = true
		if dir_x < 0 {
			event += "left"
		} else {
			event += "right"
		}
	} else if abs_y > abs_x && abs_y >= 10 {
		ok = true
		if dir_y < 0 {
			event += "up"
		} else {
			event += "down"
		}
	}
	return
}

func watch(device *evdev.InputDevice) {
	events := gatherEvents(device)
	diff := 0
	i, j := 0, 0
	dir_y, dir_x := 0, 0
	x_buf := [4]int32{}
	y_buf := [4]int32{}
	c_max := uint16(0)
	duration := 100 * time.Millisecond
	t := timer.NewTimer(duration)
	for {
		select {
		case ev := <-events:
			if ev == nil {
				break
			}
			t.Reset(duration)
			switch ev.Type {
			case evdev.EV_ABS:
				switch ev.Code {
				case evdev.ABS_X:
					diff, i = update_buff(&x_buf, ev.Value, i)
					dir_x += diff
				case evdev.ABS_Y:
					diff, j = update_buff(&y_buf, ev.Value, j)
					dir_y += diff
				}
			case evdev.EV_KEY:
				if ev.Code == evdev.BTN_TOOL_QUADTAP || ev.Code == evdev.BTN_TOOL_TRIPLETAP {
					if ev.Code > c_max {
						c_max = ev.Code
					}
				}
			}
		case <-t.C:
			event, ok := get_gesture_name(dir_x, dir_y, c_max)
			if ok {
				fmt.Println(event)
			}
			i, j = 0, 0
			dir_x, dir_y = 0, 0
			c_max = 0
		}
	}
}

func main() {
	if len(os.Args) == 1 {
		fmt.Println(`usage:
	midas list
	midas <name>
	midas <path>`)
		os.Exit(1)
	}
	if os.Args[1] == "list" {
		devices, _ := evdev.ListInputDevices()
		for _, device := range devices {
			fmt.Println(device.File.Name(), device.Name)
		}
	} else if len(os.Args[1]) > 0 && os.Args[1][0] == '/' {
		device, err := evdev.Open(os.Args[1])
		if err != nil {
			_, _ = fmt.Fprintf(os.Stderr, "midas: %e\n", err)
		}
		watch(device)
	} else {
		devices, _ := evdev.ListInputDevices()
		for _, device := range devices {
			if device.Name == os.Args[1] {
				watch(device)
				break
			}
		}
	}
}
